src/parser.ml

```diff
@@ -967,28 +967,61 @@ module Pre = struct
+                && (n1 + n2) mod 3 = 0
+                && n1 mod 3 != 0
+                && n2 mod 3 != 0
+              then
+                let rec find_next_emph = function
+                  | Emph (_, _, _, n) :: _ -> Some n
+                  | _ :: xs -> find_next_emph xs
+                  | [] -> None
+                in
+                let next_emph = find_next_emph xs in
+                match next_emph with
+                | None -> loop (x :: acc) xs
+                | Some n ->
+                    if (n + n2) mod 3 = 0 && n mod 3 != 0 && n2 mod 3 != 0 then
+                      loop (x :: acc) xs
+                    else
+                      let xs' = parse_emph xall in
+                      loop acc xs'
+              else
+                let xs =
+                  if n1 >= 2 && n2 >= 2 then
+                    if n2 > 2 then Emph (Other, post, q2, n2 - 2) :: xs else xs
+                  else if n2 > 1 then Emph (Punct, post, q2, n2 - 1) :: xs
+                  else xs
+                in
+                let r =
+                  let il = concat (List.map to_r (parse_emph (List.rev acc))) in
+                  if n1 >= 2 && n2 >= 2 then R (Strong ([], il)) :: xs
+                  else R (Emph ([], il)) :: xs
+                in
+                let r =
+                  if n1 >= 2 && n2 >= 2 then
+                    if n1 > 2 then Emph (pre, Other, q1, n1 - 2) :: r else r
+                  else if n1 > 1 then Emph (pre, Punct, q1, n1 - 1) :: r
+                  else r
+                in
+                parse_emph r
+          | (Emph (_, _, q2, _) as x) :: xs1 as xs when is_opener x ->
+              let rec find_next_closer_emph = function
+                | (Emph (_, _, q, _) as x) :: _ when is_closer x -> Some q
+                | _ :: xs -> find_next_closer_emph xs
+                | [] -> None
               in
-              let r =
-                let il = concat (List.map to_r (parse_emph (List.rev acc))) in
-                if n1 >= 2 && n2 >= 2 then R (Strong ([], il)) :: xs
-                else R (Emph ([], il)) :: xs
-              in
-              let r =
-                if n1 >= 2 && n2 >= 2 then
-                  if n1 > 2 then Emph (pre, Punct, q1, n1 - 2) :: r else r
-                else if n1 > 1 then Emph (pre, Punct, q1, n1 - 1) :: r
-                else r
+              let next_closer_emph = find_next_closer_emph xs1 in
+              let is_next_closer_same =
+                match next_closer_emph with None -> false | Some q3 -> q2 = q3
               in
-              parse_emph r
-          | (Emph _ as x) :: xs1 as xs when is_opener x ->
-              let xs' = parse_emph xs in
-              if xs' = xs then loop (x :: acc) xs1 else loop acc xs'
+              if not is_next_closer_same then loop (x :: acc) xs1
+              else
+                let xs' = parse_emph xs in
+                if xs' = xs then loop (x :: acc) xs1 else loop acc xs'
```

Comment:
I haven't puzzled out this function enough, do you happen to know why we are doing a (polymorphic) equality comparison on the lists here

Reply:
I have no idea 😬 I could investigate but that comparison doesn't seem to be needed as the following code successfully pass all the tests.

```diff
diff --git a/src/parser.ml b/src/parser.ml
index e90b96d..46f37b6 100644
--- a/src/parser.ml
+++ b/src/parser.ml
@@ -959,9 +959,7 @@ module Pre = struct
                 | Some (_, _, q3, _) -> q2 = q3
               in
               if not is_next_closer_same then loop (x :: acc) xs1
-              else
-                let xs' = parse_emph xs in
-                if xs' = xs then loop (x :: acc) xs1 else loop acc xs'
+              else loop acc (parse_emph xs)
           | x :: xs -> loop (x :: acc) xs
           | [] -> x :: List.rev acc
         in

```

Reply:
That's already a good investigation! I'll make note to followup here with a change that removes this bit. Thanks!

Reply:
I tried removing that equality comparison in master

```diff
diff --git a/src/parser.ml b/src/parser.ml
index 6342989..053ca50 100644
--- a/src/parser.ml
+++ b/src/parser.ml
@@ -926,9 +926,8 @@ module Pre = struct
                 else r
               in
               parse_emph r
-          | (Emph _ as x) :: xs1 as xs when is_opener x ->
-              let xs' = parse_emph xs in
-              if xs' = xs then loop (x :: acc) xs1 else loop acc xs'
+          | (Emph _ as x) :: _ as xs when is_opener x ->
+              loop acc (parse_emph xs)
           | x :: xs -> loop (x :: acc) xs
           | [] -> x :: List.rev acc
         in
```

When running the tests, it gets stuck at 99%. I assume it's because of an infinite loop. So it appears that this code is useful in master but becomes unnecessary with that PR.